frontends/pytorch/csrc/builder/ivalue_importer.cpp

```diff
@@ -0,0 +1,177 @@
+//===- ivalue_importer.cpp ------------------------------------------------===//
+//
+// This file is licensed under a pytorch-style license
+// See frontends/pytorch/LICENSE for license information.
+//
+//===----------------------------------------------------------------------===//
+
+#include "ivalue_importer.h"
+#include "graph_importer.h"
+
+#include <unordered_map>
+
+#include "mlir_utils.h"
+
+#include "mlir-c/BuiltinAttributes.h"
+#include "mlir-c/BuiltinTypes.h"
+#include "mlir-c/Diagnostics.h"
+#include "npcomp-c/Types.h"
+
+using namespace torch_mlir;
+
+namespace {
+/// Helper class for holding state during recursive IValue import.
+///
+/// The intended usage pattern of this class is to construct it then call
+/// `importIValue` exactly once. Calling `importIValue` more than once
+/// is likely to produce unexpected results since the same in-memory IValue
+/// can be reimported more than once. That is, subsequent calls to
+/// `importIValue` will not properly unify IValue's with already-imported
+/// IValue's.
+///
+/// TODO: Support unifying repeated IValue's.
+/// This already is an issue when importing a single IValue through the current
+/// API, because the same underlying Tensor (or List/Dict) can be referenced by
+/// multiple properties of a module. There is an extra complication with tensors
+/// because they can alias each other in fairly arbitrary ways, which we will
+/// need to model with slice ops.
+class IValueImporter {
+public:
+  IValueImporter(MlirBlock importBlock, MlirContext context)
+      : importBlock(importBlock), context(context), typeMapper(context) {}
+
+  MlirValue importIValue(c10::IValue value);
+
+private:
+  MlirValue importModule(torch::jit::Module jitModule);
+  void importMethod(torch::jit::Function *function, MlirBlock nnModuleBody);
+
+  MlirBlock importBlock;
+  MlirContext context;
+  TypeMapper typeMapper;
+};
+} // namespace
+
+MlirValue IValueImporter::importModule(torch::jit::Module currentModule) {
+  // TODO: Can we do better?
```

Comment:
See GraphImporter::extractCallstackLoc(). Probably needs to be lifted to your new common utils. Fine to do in a follow-up.

Also, I've found that I've had to fiddle with this in the past. Not all nodes in TorchScript have sensical locations and it required some probing.

Reply:
The problem is that torch::jit::Module's are not Node's. I guess I could pick a random Node from one of the methods of the module (if there even is a method) and at least get a file name out of it. That seems unsatisfactory though. I would prefer to leave it as-is and wait until we see what kinds of diagnostics we need to emit on these to see what is useful.